File: frontend/lib/models/meeting.dart
import 'package:json_annotation/json_annotation.dart';

part 'meeting.g.dart';

@JsonSerializable()
class Meeting {
  final int id;
  final String title;
  final String? description;
  final String audioFilePath;
  final double? audioDuration;
  final String? transcription;
  final List<TranscriptionSegment>? transcriptionSegments;
  final String? detectedLanguage;
  final String? modelUsed;
  final String? summary;
  final List<String>? actionItems;
  final List<String>? keyPoints;
  final String status;
  final String? errorMessage;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? processedAt;

  Meeting({
    required this.id,
    required this.title,
    this.description,
    required this.audioFilePath,
    this.audioDuration,
    this.transcription,
    this.transcriptionSegments,
    this.detectedLanguage,
    this.modelUsed,
    this.summary,
    this.actionItems,
    this.keyPoints,
    required this.status,
    this.errorMessage,
    required this.createdAt,
    this.updatedAt,
    this.processedAt,
  });

  factory Meeting.fromJson(Map<String, dynamic> json) => _$MeetingFromJson(json);
  Map<String, dynamic> toJson() => _$MeetingToJson(this);

  String get formattedDuration {
    if (audioDuration == null) return 'Unknown';
    final minutes = (audioDuration! / 60).floor();
    final seconds = (audioDuration! % 60).floor();
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }

  bool get isProcessing => status == 'processing';
  bool get isCompleted => status == 'completed';
  bool get isFailed => status == 'failed';
}

@JsonSerializable()
class TranscriptionSegment {
  final double start;
  final double end;
  final String text;

  TranscriptionSegment({
    required this.start,
    required this.end,
    required this.text,
  });

  factory TranscriptionSegment.fromJson(Map<String, dynamic> json) =>
      _$TranscriptionSegmentFromJson(json);
  Map<String, dynamic> toJson() => _$TranscriptionSegmentToJson(this);

  String get timeRange {
    final startMin = (start / 60).floor();
    final startSec = (start % 60).floor();
    final endMin = (end / 60).floor();
    final endSec = (end % 60).floor();
    return '$startMin:${startSec.toString().padLeft(2, '0')} - $endMin:${endSec.toString().padLeft(2, '0')}';
  }
}

File: backend/app/models/meeting.py
from sqlalchemy import Column, Integer, String, Text, DateTime, Float, Boolean, JSON
from sqlalchemy.sql import func
from app.models import Base


class Meeting(Base):
    __tablename__ = "meetings"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    audio_file_path = Column(String(512), nullable=False)
    audio_duration = Column(Float, nullable=True)  # in seconds

    # Transcription
    transcription = Column(Text, nullable=True)
    transcription_segments = Column(JSON, nullable=True)  # Timestamped segments
    detected_language = Column(String(10), nullable=True)
    model_used = Column(String(100), nullable=True)  # whisper or wav2vec

    # LLM Processing
    summary = Column(Text, nullable=True)
    action_items = Column(JSON, nullable=True)  # List of action items
    key_points = Column(JSON, nullable=True)  # List of key discussion points

    # Status
    status = Column(String(50), default="uploaded")  # uploaded, processing, completed, failed
    error_message = Column(Text, nullable=True)

    # Metadata
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    processed_at = Column(DateTime(timezone=True), nullable=True)

    def __repr__(self):
        return f"<Meeting(id={self.id}, title='{self.title}', status='{self.status}')>"
